src/core/cgroup.c

```diff
@@ -181,6 +182,270 @@ void cgroup_context_init(CGroupContext *c) {
         };
 }
 
+int cgroup_context_copy(const CGroupContext *src, CGroupContext *dst) {
```

Comment:
uh, I don#t like this at all. If people want to enforce limits on service units and their auxiliary units they should set this on a surrounding slice. it's simply wrong to copy the values here, since it will come from different pools.

Sorry, but this part really should go. focus for inheriting resource mgmt must be on inheriting the parent slice, not on inheriting settings.

Reply:
To be frank, I don't like that part either because it is ugly but I think this is necessary. Otherwise we basically make this feature unusable for anything that runs under global mostly unrestricted slice, e.g. system.slice which are *all* services on default distro installations. 

On top of that, failure to setup proper slice hierarchy and managing resources and *security* strictly on slice level could lead to unintended consequences. For example by using this feature you could allow access to certain hosts that previously the processes did not have access to.

In other words, I don't think we should be forcing people to use slice hierarchy. Btw, I am not against it in general, I think it is a good idea but it should be a hard requirement.

Reply:
The disagreement here stems from the fact that we have two types of resource managment. The first is where we set some bounds on what a service can access an some limits on process resources in case it goes crazy, but we don't know know precisely how much resources the process will use and we don't really care. The second kind is where we control the job precisely and tailor the job to the machine resource limits. In the second case, creating a duplicate service with the same limits is going to cause problems because it'll exceed what the machine has. In the first case, the limits are just a rough assignment and doubling the values doesn't mean much. The first type is much more popular. 

I think that copying the resource limits is the expected thing in the first case. In the second case, it might not be the right thing, but then it's easy enough to adjust by placing those siblings in a slice. So overall, copying resource limits makes that casual case easy, and doesn't make the second case, with precise resource allocations, any worse. In fact, it's necessary even in the second case, because by default we want equal treating of both instances, which means we should have the same settings on them, so that whatever the parent slice is given is fairly divided between the children.

Overall, I don't find the routine to copy settings nice at all, but it's also rather simple, even if long, and should be easy to maintain. Only when we add new unit properties, we have to modify the function. This doesn't happen too often, so I don't think this will create any maintenance problems.

Reply:
> Overall, I don't find the routine to copy settings nice at all, but it's also rather simple, even if long, and should be easy to maintain. Only when we add new unit properties, we have to modify the function. This doesn't happen too often, so I don't think this will create any maintenance problems.

Yeah I agree - also normally one would configure the scope with the required limits, but this is an automated way to migrate processes to a new implicit scope, and the closest thing is to clone the existing settings of the original unit. In fact, it would be quite surprising to me if processes were under specific limits under the original unit (eg, a certain MemoryHigh to put them under pressure) but were completely unrestricted once migrated, without manually configuring another slice on the fly before asking for the migration.